pkg/inctimer/inctimer.go

```diff
@@ -0,0 +1,47 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package inctimer
+
+import "time"
+
+// IncTimer wraps time.Timer
+type IncTimer struct {
+	t *time.Timer
+}
+
+// New creates a new IncTimer and a done function
+func New() (*IncTimer, func() bool) {
+	t := time.NewTimer(time.Nanosecond)
+	return &IncTimer{
+		t: t,
+	}, t.Stop
+}
+
+// After returns a channel that will fire after
+// the specified duration.
+func (it *IncTimer) After(d time.Duration) <-chan time.Time {
+	// We cannot call reset on an expired timer,
+	// so we need to stop it and drain it first.
+	// See https://golang.org/pkg/time/#Timer.Reset for more details.
+	if !it.t.Stop() {
+		// It could be that the channel was read already
+		select {
+		case <-it.t.C:
+		default:
+		}
+	}
+	it.t.Reset(d)
+	return it.t.C
+}
```

Comment:
I'm not sure why we need a wrapper. Why isn't the fix mentioned in the medium article sufficient?

Reply:
+1

Reply:
The solution suggested in the article is actually partially incorrect. In order to reset a timer  safely (as suggested in the article), you have to also [stop the timer and clear the channel](https://golang.org/pkg/time/#Timer.Reset).
I think this wrapper is useful because it basically boils the following code:
```golang
t := time.NewTimer(time.Nanosecond)
defer t.Stop()
for {
...
        if !t.Stop() {
		// It could be that the channel was read already
		select {
		case <-t.C:
		default:
		}
	}
        t.Reset(someInterval)
        select {
        ...
        case <- t.C:
        }
}
```
to:
```golang
t, stop := inctimer.New()
defer stop()
for {
...
        select {
        ...
        case <- t.After(someInterval):
        }
}
```
In one case I believe the `After` function is called even more than once, so that would have either required the creation of two separate timers or the same hard reset logic repeated twice above the select.
I find this wrapper approach to be a lot more elegant and easier to use.